src/trusted/service_runtime/nacl_syscall_common.c

```diff
@@ -4163,24 +4132,68 @@ int32_t NaClSysExecve(struct NaClAppThread *natp, char const *path, char *const
     for (int i = 0; i < new_envc; i++) {
       char *env = (void *)NaClUserToSysAddr(nap, (uintptr_t)sys_envp_ptr[i]);
       env = (uintptr_t)env == kNaClBadAddress ? 0 : env;
-      new_envp[i] = env ? strdup(env) : 0;
       if (!env) {
+        new_envp[i] = 0;
         break;
       }
+      else {
+        int envsize = NACL_ENV_PREFIX_LENGTH + strlen(env) + 1;
+        new_envp[i] = calloc(envsize, sizeof(char));
+        snprintf(new_envp[i], envsize, "%s%s", NACL_ENV_PREFIX, env);
```

Comment:
snprintf is likely significantly slower than 2 strcpys, but I'll leave it up to you whether we care

Reply:
I had it like that initially, I think this is clearer and won't be called enough to really be a performance problem.